### PR TITLE
Handle a lost standalone ack correctly.

### LINKED ISSUE #9796: Reliable messaging crashes if a standalone ack is lost
#### Problem
Consider the following sequence of events:
1. Initiator sends a reliable message to responder.
2. Responder takes a while to process the message, hits standalone ack timeout, sends standalone ack.
3. Standalone ack is lost.
4. Responder sends an application-level response.  No ack is piggybacked, because the standalone ack was already sent.
5. Initiator processes application-level response, sends another application-level reliable message on the same exchange.

This crashes the initiator in `ReliableMessageMgr::AddToRetransTable` because of this line:

> `VerifyOrDie(rc != nullptr && !rc->IsOccupied());`

In this case, `rc->IsOccupied()` is true: it's waiting for an ack to that first message it sent.

 #### Proposed Solution
We have a few options:

* Allow a given `ReliableMessageContext` to be waiting for more than one ack.
* Don't clear ack information completely when sending a standalone ack; piggyback that same ack on the next application-level message as well.  This will require fixing the behavior we have now where processing of a message that acks something we think does not need acking is completely aborted.
* Other ideas?

@turon @kghost @yufengwangca @pan-apple @msandstedt 

<details>
<summary>Unit test exercising this scenario, which crashes on ToT:</summary>

```
diff --git a/src/messaging/tests/TestReliableMessageProtocol.cpp b/src/messaging/tests/TestReliableMessageProtocol.cpp
index 34d6ef108..ba23cd414 100644
--- a/src/messaging/tests/TestReliableMessageProtocol.cpp
+++ b/src/messaging/tests/TestReliableMessageProtocol.cpp
@@ -1111,6 +1111,135 @@ void CheckMessageAfterClosed(nlTestSuite * inSuite, void * inContext)
     NL_TEST_ASSERT(inSuite, rm->TestGetCountRetransTable() == 0);
 }
 
+void CheckLostStandaloneAck(nlTestSuite * inSuite, void * inContext)
+{
+    /**
+     * This tests the following scenario:
+     * 1) A reliable message is sent from initiator to responder.
+     * 2) The responder sends a standalone ack, which is lost.
+     * 3) The responder sends an application-level response.
+     * 4) The initiator sends a reliable response to the app-level response.
+     */
+    TestContext & ctx = *reinterpret_cast<TestContext *>(inContext);
+
+    chip::System::PacketBufferHandle buffer = chip::MessagePacketBuffer::NewWithData(PAYLOAD, sizeof(PAYLOAD));
+    NL_TEST_ASSERT(inSuite, !buffer.IsNull());
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    MockAppDelegate mockReceiver;
+    err = ctx.GetExchangeManager().RegisterUnsolicitedMessageHandlerForType(Echo::MsgType::EchoRequest, &mockReceiver);
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    mockReceiver.mTestSuite = inSuite;
+
+    MockAppDelegate mockSender;
+    ExchangeContext * exchange = ctx.NewExchangeToAlice(&mockSender);
+    NL_TEST_ASSERT(inSuite, exchange != nullptr);
+
+    mockSender.mTestSuite = inSuite;
+
+    ReliableMessageMgr * rm = ctx.GetExchangeManager().GetReliableMessageMgr();
+    NL_TEST_ASSERT(inSuite, rm != nullptr);
+
+    // Ensure the retransmit table is empty right now
+    NL_TEST_ASSERT(inSuite, rm->TestGetCountRetransTable() == 0);
+
+    // We send a message, the other side sends a standalone ack first (which is
+    // lost), then an application response, then we respond to that response.
+    // We need to keep both exchanges alive for that (so we can send the
+    // response from the receiver and so the initial sender exchange can send a
+    // response to that).
+    gLoopback.mSentMessageCount    = 0;
+    gLoopback.mNumMessagesToDrop   = 0;
+    gLoopback.mDroppedMessageCount = 0;
+    mockReceiver.mRetainExchange   = true;
+    mockSender.mRetainExchange     = true;
+
+    // And ensure the ack heading back our way is dropped.
+    mockReceiver.mDropAckResponse = true;
+
+    err = exchange->SendMessage(Echo::MsgType::EchoRequest, std::move(buffer), SendFlags(SendMessageFlags::kExpectResponse));
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Ensure the message was sent.
+    NL_TEST_ASSERT(inSuite, gLoopback.mSentMessageCount == 1);
+    NL_TEST_ASSERT(inSuite, gLoopback.mDroppedMessageCount == 0);
+
+    // And that it was received.
+    NL_TEST_ASSERT(inSuite, mockReceiver.IsOnMessageReceivedCalled);
+
+    // And that we have not gotten any app-level responses or acks so far.
+    NL_TEST_ASSERT(inSuite, !mockSender.IsOnMessageReceivedCalled);
+    NL_TEST_ASSERT(inSuite, rm->TestGetCountRetransTable() == 1);
+
+    ReliableMessageContext * receiverRc = mockReceiver.mExchange->GetReliableMessageContext();
+    // Ack should have been dropped.
+    NL_TEST_ASSERT(inSuite, !receiverRc->IsAckPending());
+
+    // Don't drop any more acks.
+    mockReceiver.mDropAckResponse = false;
+
+    // Now send a message from the other side.
+    buffer = chip::MessagePacketBuffer::NewWithData(PAYLOAD, sizeof(PAYLOAD));
+    NL_TEST_ASSERT(inSuite, !buffer.IsNull());
+
+    mockReceiver.mExchange->SendMessage(Echo::MsgType::EchoResponse, std::move(buffer),
+                                        SendFlags(SendMessageFlags::kExpectResponse));
+
+    // Ensure the response was sent.
+    NL_TEST_ASSERT(inSuite, gLoopback.mSentMessageCount == 2);
+    NL_TEST_ASSERT(inSuite, gLoopback.mDroppedMessageCount == 0);
+
+    // Ensure that we have received that response and it did not have a
+    // piggyback ack.
+    NL_TEST_ASSERT(inSuite, mockSender.IsOnMessageReceivedCalled);
+    NL_TEST_ASSERT(inSuite, !mockSender.mReceivedPiggybackAck);
+    // We now have our un-acked message still waiting to retransmit and the
+    // message we just received is waiting for an ack.
+    NL_TEST_ASSERT(inSuite, rm->TestGetCountRetransTable() == 2);
+
+    // Reset various state so we can measure things again.
+    mockReceiver.IsOnMessageReceivedCalled = false;
+    mockReceiver.mReceivedPiggybackAck     = false;
+    mockSender.IsOnMessageReceivedCalled   = false;
+    mockSender.mReceivedPiggybackAck       = false;
+
+    // Stop retaining the recipient exchange.
+    mockReceiver.mRetainExchange = false;
+
+    // Now send a new message to the other side.
+    buffer = chip::MessagePacketBuffer::NewWithData(PAYLOAD, sizeof(PAYLOAD));
+    NL_TEST_ASSERT(inSuite, !buffer.IsNull());
+
+    // Claim (falsely) that we expect a response, so exchange stays open and we
+    // can use it to clear out reliable message state later.
+    err = exchange->SendMessage(Echo::MsgType::EchoRequest, std::move(buffer),
+                                SendFlags(SendMessageFlags::kExpectResponse));
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+
+    // Ensure the message and the standalone ack to it were sent.
+    NL_TEST_ASSERT(inSuite, gLoopback.mSentMessageCount == 4);
+    NL_TEST_ASSERT(inSuite, gLoopback.mDroppedMessageCount == 0);
+
+    // And that it was received.
+    NL_TEST_ASSERT(inSuite, mockReceiver.IsOnMessageReceivedCalled);
+    NL_TEST_ASSERT(inSuite, mockReceiver.mReceivedPiggybackAck);
+
+    // And that receiver has no ack pending.
+    NL_TEST_ASSERT(inSuite, !receiverRc->IsAckPending());
+
+    // And that we are still expecting an ack for the one message an ack was
+    // lost for.
+    NL_TEST_ASSERT(inSuite, rm->TestGetCountRetransTable() == 1);
+
+    // Clear out our one pending retransmission entry.
+    rm->ClearRetransTable(exchange);
+
+    // And we should have no pending work left now.
+    NL_TEST_ASSERT(inSuite, rm->TestGetCountRetransTable() == 0);
+}
+
 // Test Suite
 
 /**
@@ -1134,6 +1263,7 @@ const nlTest sTests[] =
     NL_TEST_DEF("Test ReliableMessageMgr::CheckSendStandaloneAckMessage", CheckSendStandaloneAckMessage),
     NL_TEST_DEF("Test command, response, default response, with receiver closing exchange after sending response", CheckMessageAfterClosed),
     NL_TEST_DEF("Test that unencrypted message is dropped if exchange requires encryption", CheckUnencryptedMessageReceiveFailure),
+    NL_TEST_DEF("Test that an application-level response-to-response after a lost standalone ack to the initial message works", CheckLostStandaloneAck),
 
     NL_TEST_SENTINEL()
 };
```

</details>

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
If a standalone ack gets lost and then there is an application-level
reply to the message that triggered the standalone ack, we want to
piggyback an ack on that application-level reply as well.  Otherwise
the other side can end up in a state where it has two outstanding
unacknowledged messages (the message we are replying to, and the reply
to our reply), which it can't handle.

The other change here is to fix ReliableMessageMgr to not crash if the
peer misbehaves and skips sending that piggyback ack described above.
Instead, we just error out from sending the response to the peer's
broken message.

Fixes https://github.com/project-chip/connectedhomeip/issues/9796

#### Problem
See above.

#### Change overview
See above.

#### Testing
New unit test passes.

### CODE REVIEW

File: src/messaging/ExchangeMessageDispatch.cpp [hunk 1]
```diff
@@ -50,7 +50,7 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SessionHandle session, uint16_t
     payloadHeader.SetExchangeID(exchangeId).SetMessageType(protocol, type).SetInitiator(isInitiator);
 
     // If there is a pending acknowledgment piggyback it on this message.
-    if (reliableMessageContext->IsAckPending())
+    if (reliableMessageContext->ShouldPiggybackAck())
```
Comment (tcarmelveilleux) [inline src/messaging/ExchangeMessageDispatch.cpp:53]: Rename to `HasPiggybackAckPending`?

File: src/messaging/ReliableMessageContext.cpp [hunk 2]
```diff
@@ -296,6 +305,7 @@ void ReliableMessageContext::SetPendingPeerAckMessageCounter(uint32_t aPeerAckMe
 {
     mPendingPeerAckMessageCounter = aPeerAckMessageCounter;
     SetAckPending(true);
+    mFlags.Set(Flags::kFlagHasHadAckPending);
```
Comment (tcarmelveilleux) [inline src/messaging/ReliableMessageContext.cpp:308]: "HasHad" in the name makes it hard to parse. I was at first wondering if SetAckPending(true) should have done it, but the answer is probably no. I am struggling, looking at the names of the methods and the flow, to really figure-out if anything is missing or going awry. There is a bit of state combinatorial explosion with different cases... Could `kFlagHasHadAckPending` be renamed to something grammatically less complex?

File: src/messaging/ReliableMessageContext.h [hunk 3]
```diff
@@ -200,16 +208,22 @@ class ReliableMessageContext
...
+        /// some message we have needed to acknowledge in the past.
+        kFlagHasHadAckPending = 0x0040,
+
         /// When set, signifies that at least one message has been received from peer on this exchange context.
-        kFlagMsgRcvdFromPeer = 0x0040,
+        kFlagMsgRcvdFromPeer = 0x0080,
 
         /// When set, signifies that this exchange is waiting for a call to SendMessage.
-        kFlagWillSendMessage = 0x0080,
+        kFlagWillSendMessage = 0x00100,
```
Comment (tcarmelveilleux) [inline src/messaging/ReliableMessageContext.h:221]: Nit: One too many leading zeros. Suggest `0x0100`.

Overall if you are shifting all these items, perhaps it would be better with (1u << 0), (1u << 1), ... (1u << 8) ,etc

File: src/messaging/ReliableMessageMgr.cpp [hunk 4]
```diff
@@ -267,7 +267,16 @@ CHIP_ERROR ReliableMessageMgr::AddToRetransTable(ReliableMessageContext * rc, Re
     bool added     = false;
     CHIP_ERROR err = CHIP_NO_ERROR;
 
-    VerifyOrDie(rc != nullptr && !rc->IsOccupied());
+    VerifyOrDie(rc != nullptr);
+
+    if (rc->IsOccupied())
+    {
+        // This can happen if we have a misbehaving peer that is not sending
+        // acks with its application-level responses when it should, so we end
```
Comment (tcarmelveilleux) [inline src/messaging/ReliableMessageMgr.cpp:275]: What does "when it should" mean? Piggybacking acks is not a deterministic operation.

File: src/messaging/tests/TestReliableMessageProtocol.cpp [hunk 5]
```diff
@@ -1246,6 +1245,123 @@ void CheckLostResponseWithPiggyback(nlTestSuite * inSuite, void * inContext)
     NL_TEST_ASSERT(inSuite, rm->TestGetCountRetransTable() == 0);
 }
 
+void CheckLostStandaloneAck(nlTestSuite * inSuite, void * inContext)
+{
+    /**
+     * This tests the following scenario:
```
Comment (tcarmelveilleux) [inline src/messaging/tests/TestReliableMessageProtocol.cpp:1251]: The expected outcome is not made clear.

File: src/messaging/ReliableMessageContext.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageContext.cpp:308]: How about `kFlagAckMessageCounterIsValid` ?  @tcarmelveilleux

File: src/messaging/ReliableMessageContext.h [hunk 6]
```diff
@@ -58,16 +58,24 @@ class ReliableMessageContext
     CHIP_ERROR FlushAcks();
 
     /**
-     * Take the pending peer ack message counter from the context.  This must only be called
-     * when IsAckPending() is true.  After this call, IsAckPending() will be
-     * false; it's the caller's responsibility to send the ack.
+     * Take the pending peer ack message counter from the context.  This must
+     * only be called when ShouldPiggybackAck() is true.  After this call,
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageContext.h:62]: Using `assert`?  My understanding is that we are avoiding using that, generally, but maybe I misunderstand the situation?

File: src/messaging/ReliableMessageMgr.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageMgr.cpp:275]: It actually needs to be.  Specifically, an application-level response to a message needs to carry the piggyback ack for that message, if the message was sent using MRP.  That was the outcome of the discussion in #9796.

But now that you make me think through this again, I think things are still broken even with the changes here.  Consider this sequence of events, all on the same exchange; numbering only applies to app-level messages:
1. A sends message 1 to B
2. B is slow to respond, A does a resend and the resend is delayed in the network.
3. B responds with message 2, which acks message 1.
4. A sends message 3
5. B sends standalone ack to message 3, which is lost
6. The duplicate message from step 3 is delivered and triggers a standalone ack.
7. B responds with message 4, which carries a piggyback ack... for message 1 (!)
8. A tries to send message 5, but still has an un-acked message 3, and things fail.

I will fix this, but it really seems like we need a spec-level issue to sort this out.  And I don't see a good way to write a test for this situation with our current test infrastructure.  :( @turon @msandstedt @kghost

Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageMgr.cpp:275]: Filed https://github.com/CHIP-Specifications/connectedhomeip-spec/issues/4168 on the spec.

File: src/messaging/ReliableMessageContext.h (hunk 3 above)
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageContext.h:221]: Done.

File: src/messaging/tests/TestReliableMessageProtocol.cpp (hunk 5 above)
Comment (bzbarsky-apple) [reply src/messaging/tests/TestReliableMessageProtocol.cpp:1251]: Added comments.

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by kghost** [comment]
I feel that `IsAckPending` is indicating the active state of an exchange. It means that the exchange is busy generating an response, and it is unable to process incoming requests now.

`HasPiggybackAckPending` means that a ack is set, it actually means that the exchange have received a packet from peer. After this PR, all packets except the very beginning one will contains a valid ack.

So I just feel that these function names are a little bit confused.

**Comment by bzbarsky-apple** [comment]
@kghost Please suggest what you'd prefer as names?

File: src/messaging/ReliableMessageMgr.cpp (hunk 4 above)
Comment (msandstedt) [reply src/messaging/ReliableMessageMgr.cpp:275]: Thanks.  I'll take my comments to #4168.

**Review by kghost — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]
